mongorestore/filepath.go

```diff
@@ -204,13 +204,17 @@ func (f *stdinFile) Close() error {
 	return nil
 }
 
-// getInfoFromFilename pulls the base collection name and FileType from a given file.
-func (restore *MongoRestore) getInfoFromFilename(filename string) (string, FileType, error) {
+// getCollectionNameFromFile returns the collection name and FileType for a bson or metadata file.
+// The collection name may be pulled from either the file name itself, or from the content of a
+// .metadata.json file if the file name is truncated.
+func (restore *MongoRestore) getCollectionNameFromFile(filename string) (string, FileType, error) {
```

Comment:
I have a couple thoughts and questions about this method and its new helpers.

1. The `getCollectionNameFromMetadata` and `getCollectionNameFromBson` methods are very similarly implemented so I wonder if that's the right breakdown of helper methods. The [detailed design section](https://docs.google.com/document/d/1LCsXHqjQrRWDzD1Of_pNX41pyQZqd231ni9uF9_noY8/edit#heading=h.ddsw2s91f9e6) actually suggests always checking the metadata file no matter what. If you fail to find the collection name by doing that, then return an error if the `filename` contains `"%24"`; otherwise you can trim the suffix and what's left is the collection name. This implies breaking down the helpers as `getCollectionNameFromMetadata` and `getCollectionNameFromFilename`. The former would need to ensure the `.metadata.json[.gz]` file exists, can be opened, and contains a `collectionName` field. The latter would need to strip away the suffix. Notice that there is no overlap of functionality between those two, whereas the current two helpers have a lot of the same code duplicated.
2. Given that, the body of this method doesn't exactly suit what we need to do now. I think we can slightly change the suggestion in the design doc to limit the number of places we call helper methods:
	1. Get the suffix and `FileType` from the `filename`. This can be done with the same conditional structure that already exists, just alter the bodies to store the `FileType` and suffix instead of calling helper functions.
	2. If the `filename` contains `"%24"`, get the collection name from the metadata file (`getColllectionNameFromMetadata`). Return an error if we can't get it for some reason.
	3. Otherwise, get the collection name from the file name by trimming the suffix (`getCollectionNameFromFilename`).
	4. Unescape (i.e. the last 6 lines of this method don't need to change).

Reply:
(This is just trying to cut back on the duplicated code between the two helpers, but that is not ultimately very important!)

Reply:
I tend to agree with your points. If we move the `%24` check up, it can definitely eliminate some duplicated code since we know we'll be looking at a metadata file next.

As a side note, I might've separated them out because I had this idea that they could also be used directly as non-helpers (e.g. if you're in a function that only expects metadata files). There's currently barely any use cases like that though, and likewise I think it's better to keep it compact than offer functions that don't really accomplish anything different than what `getCollectionNameFromFilename` accomplishes.

Reply:
I agree with what Matt laid out. Although the design doc says to always check the metadata file for the collection name, that's inefficient and I think we should only do that if %24 is in the filename.

I don't agree with renaming this function from `getInfoFromFilename` to `getCollectionNameFromFile` by the way. It returns more than just a filename.